File: Pascals Triangle 2/pascals_triangle.py
'''
https://leetcode.com/problems/pascals-triangle-ii/
time complexity: O(rowIndex * rowIndex)
space complexity: O(rowIndex)
'''
class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        if rowIndex == 0:
            return [1]
        if rowIndex == 1:
            return [1,1]
        prev = [1,1]
        for i in range(2,rowIndex+1):
            temp = [1]
            len_prev = len(prev)
            for j in range(1,len_prev):
                temp.append(prev[j-1]+prev[j])
            temp.append(1)
            prev = temp
        return prev
        

File: Pascals Triangle 2/pascalsTriangle.java
import java.util.ArrayList;
import java.util.List;

/**
time complexity: O(rowIndex * rowIndex)
space complexity: O(rowIndex)
 */
class Solution {
    public List<Integer> getRow(int rowIndex) {
        if(rowIndex == 0) return List.of(1);
        if(rowIndex == 1) return List.of(1,1);
        List<Integer> prev = List.of(1,1);
        for(int i=2;i<rowIndex+1;i++)
        {
            List<Integer> temp = new ArrayList<>();
            temp.add(1);
            int lenPrev = prev.size();
            for(int j = 1;j<lenPrev;j++)
            {
                temp.add(prev.get(j-1)+prev.get(j));
            }
            temp.add(1);
            prev = temp;
        }
        return prev;
    }
}